### PR TITLE
make lt-comp go a bit faster

### PR DESCRIPTION
**Pull request by mr-martian**
Compiling the cat monodix was taking about 15 seconds that seemed kind of high to me so I implemented the changes mentioned in #96 and also poked at some other things and managed to get the cat compile time down to about 12.5 seconds.

From profiling it looks like 80% of the runtime of `lt-comp` is in `transducer.minimize()` which is inconveniently also the the one step of the process that I don't actually understand.

### CODE REVIEW

File: lttoolbox/compiler.cc [hunk 1]
```diff
@@ -208,13 +212,20 @@ Compiler::procParDef()
...
   {
     if(!paradigms[current_paradigm].isEmpty())
     {
-      paradigms[current_paradigm].minimize();
-      paradigms[current_paradigm].joinFinals();
+	  if (current_minimise) {
+		paradigms[current_paradigm].minimize();
+	  } else {
+		paradigms[current_paradigm].joinFinals();
+	  }
```
Comment (unhammer) [inline lttoolbox/compiler.cc:228]: Shouldn't this be `if(current_minimise) minimize(); joinFinals()` , ie. we want to joinFinals also when we minimize? (Seems like the current change would also alter how non-paradigms are compiled!)

File: lttoolbox/transducer.cc [hunk 2]
```diff
@@ -319,18 +319,18 @@ Transducer::determinize(int const epsilon_tag)
 
   int t = 0;
 
+  set<int> finals_state;
+  for(auto& it2 : finals) {
+	finals_state.insert(it2.first);
+  }
+
```
Comment (unhammer) [inline lttoolbox/transducer.cc:326]: for our edification, can you explain why this block is moved?

**Review by unhammer — COMMENTED** [review]
mind filling in the git commit messages a bit with the reasoning? (the current message just makes it seem like you fixed a missing thing in the dtd and made some internal speed optimization, but IIUC it's actually a new feature that you have to apply to your dix files to make them compile faster)

File: lttoolbox/compiler.cc (hunk 1 above)
Comment (mr-martian) [reply lttoolbox/compiler.cc:228]: `minimize()` calls `reverse()` which calls `joinFinals()`, so there shouldn't be a change in behavior.

File: lttoolbox/transducer.cc (hunk 2 above)
Comment (mr-martian) [reply lttoolbox/transducer.cc:326]: It's calculating a list of the final states in the original transducer, but in its original location it was recalculating it on each iteration but it won't have changed.

File: lttoolbox/compiler.cc (hunk 1 above)
Comment (unhammer) [reply lttoolbox/compiler.cc:228]: Calls `reverse` twice in fact, so it both joins finals and initials … I suppose nowhere in that sequence any new (unjoined) final is added so it's safe then :)

**Comment by unhammer** [comment]
LGTM

**Comment by mr-martian** [comment]
The efficiency tweaks have now been incorporated into #115, so I'm closing this since the #96 piece probably needs more discussion and merging vs redoing are likely to be of comparable effort.

**Pull request closed without merging**